### PR TITLE
Fixes Issue #302: Add XmlPoke Aliases

### PR DESCRIPTION
**Pull request by RichiCoder1**

### CODE REVIEW

**Comment by patriksvensson** [comment]
Nice! :+1: 
Will take a closer look at and merge this after the v0.5.1 release.

**Comment by ChrisMissal** [comment]
Excellent! :+1:

File: src/Cake.Common/Xml/XmlPokeAliases.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        /// Set the value of, or remove, target nodes.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="source">The source xml to transform.</param>
+        /// <param name="destination">The destination to write too.</param>
+        /// <param name="xpath">The xpath of the nodes to set.</param>
+        /// <param name="value">The value to set too. Leave blank to remove the selected nodes.</param>
+        /// <param name="settings">Additional settings to tweak Xml Poke behavior.</param>
+        [CakeMethodAlias]
+        public static void XmlPoke(this ICakeContext context, XmlReader source, XmlWriter destination, string xpath, string value, XmlPokeSettings settings)
```
Comment (patriksvensson) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: Don't see much reason to expose this. I suspect the two above methods will suffice.

File: src/Cake.Common/Xml/XmlPokeSettings.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+{
+    /// <summary>
+    /// Contains settings for <see cref="XmlPokeAliases"/>
+    /// </summary>
+    public sealed class XmlPokeSettings
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="XmlPokeSettings"/> class.
+        /// </summary>
+        public XmlPokeSettings()
```
Comment (patriksvensson) [inline src/Cake.Common/Xml/XmlPokeSettings.cs]: Properties before constructor.

File: src/Cake.Common/Xml/XmlPokeAliases.cs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    {
+        /// <summary>
+        /// Set the value of, or remove, target nodes.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="filePath">The target file.</param>
+        /// <param name="xpath">The xpath of the nodes to set.</param>
+        /// <param name="value">The value to set too. Leave blank to remove the selected nodes.</param>
+        /// <param name="settings">Additional settings to tweak Xml Poke behavior.</param>
+        [CakeMethodAlias]
```
Comment (patriksvensson) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: Would be great if we could add an example of how to use the alias with the `example` XML tag.

File: src/Cake.Common/Xml/XmlPokeAliases.cs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        /// Set the value of, or remove, target nodes.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="sourceXml">The source xml to transform.</param>
+        /// <param name="xpath">The xpath of the nodes to set.</param>
+        /// <param name="value">The value to set too. Leave blank to remove the selected nodes.</param>
+        /// <param name="settings">Additional settings to tweak Xml Poke behavior.</param>
+        /// <returns>Resulting XML.</returns>
+        [CakeMethodAlias]
+        public static string XmlPoke(this ICakeContext context, string sourceXml, string xpath, string value, XmlPokeSettings settings = null)
```
Comment (patriksvensson) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: Would be great if we could add an example of how to use the alias with the `example` XML tag.

File: src/Cake.Common/Xml/XmlPokeSettings.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+        /// <summary>
+        /// Gets or sets a value indicating whether to preserve white space.
+        /// </summary>
+        public bool PreserveWhitespace { get; set; }
+
+        /// <summary>
+        /// Gets or sets namespaces to include for xpath recognition.
+        /// </summary>
+        public Dictionary<string, string> Namespaces { get; set; }
```
Comment (patriksvensson) [inline src/Cake.Common/Xml/XmlPokeSettings.cs]: Should probably be `IDictionary<string, string>` to follow other tool alias.

**Comment by patriksvensson** [comment]
@RichiCoder1 I've taken a look at it now and it looks good.

I left some comments about some thoughts I had. Minor stuff :smile:

File: src/Cake.Common/Xml/XmlPokeAliases.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+            if (settings == null)
+            {
+                settings = new XmlPokeSettings();
+            }
+
+            using (var resultStream = new MemoryStream())
+            using (var sourceReader = new StringReader(sourceXml))
+            {
+                var xmlReader = XmlReader.Create(sourceReader);
```
Comment (devlead) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: I believe `XmlReader` & `XmlWriter` inherit from `IDisposable` and should be wrapped in a using statement.

Comment (RichiCoder1) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: I thought so too, but it looks like that's redundant, as the Dispose on those two just clean up the underlying streams (which we're already doing).

File: src/Cake.Common/Xml/XmlPokeAliases.cs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                settings = new XmlPokeSettings();
+            }
+
+            var sourceFile = context.FileSystem.GetFile(filePath);
+            if (!sourceFile.Exists)
+            {
+                throw new FileNotFoundException("Source File not found.", sourceFile.Path.FullPath);
+            }
+            
+            using (var sourceReader = sourceFile.Open(FileMode.Open, FileAccess.ReadWrite, FileShare.None))
```
Comment (devlead) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: Minor as `sourceFile`& `sourceReader` used for both read&write their naming might be a bit misleading, perhaps `file` & `fileStream` is more appropriate.

File: src/Cake.sln.DotSettings [hunk 8]
```diff
@@ -15,6 +15,7 @@
 	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/Abbreviations/=NET/@EntryIndexedValue">NET</s:String>
 	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/Abbreviations/=VS/@EntryIndexedValue">VS</s:String>
 	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateStaticReadonly/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="_" Suffix="" Style="aaBb" /&gt;</s:String>
+	<s:String x:Key="/Default/CodeStyle/TypeScriptCodeStyle/ImportReferenceStyle/@EntryValue">ECMAScript6</s:String>
```
Comment (devlead) [inline src/Cake.sln.DotSettings]: Don't think this file is necessary for this PR.

File: src/Cake.Common/Xml/XmlPokeSettings.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    /// </summary>
+    public sealed class XmlPokeSettings
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="XmlPokeSettings"/> class.
+        /// </summary>
+        public XmlPokeSettings()
+        {
+            Namespaces = new Dictionary<string, string>();
+            Encoding = Encoding.Default;
```
Comment (devlead) [inline src/Cake.Common/Xml/XmlPokeSettings.cs]: As the most common encoding is `UTF8` and Default is unknown (based on system settings) I think `UTF8` should be the "default".

**Comment by devlead** [comment]
Looks ok to me, left some minor comments and agree with the issues @patriksvensson raised.

File: src/Cake.Common/Xml/XmlPokeAliases.cs (hunk 6 above)
Comment (devlead) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: Ok, just checked the docs which stated `Closes the XmlWriter and the underlying stream/TextReader (if Settings.CloseOutput is true).` which led me to believe it did something more. Haven't checked IL so might be as you say, if so it's a non issue.

**Comment by devlead** [comment]
Also would be nice with at least one test case to ensure expected behavior and ti have as an sample of usage,

**Comment by patriksvensson** [comment]
@RichiCoder1 i'm planning on making a v0.5.2 release tomorrow (or later today). Should I move this to the v0.5.3 milestone?

File: src/Cake.Common/Xml/XmlPokeAliases.cs (hunk 1 above)
Comment (RichiCoder1) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: Will do.

File: src/Cake.Common/Xml/XmlPokeAliases.cs (hunk 3 above)
Comment (RichiCoder1) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: Wil do! Will add.

File: src/Cake.Common/Xml/XmlPokeSettings.cs (hunk 5 above)
Comment (RichiCoder1) [inline src/Cake.Common/Xml/XmlPokeSettings.cs]: Can do.

File: src/Cake.Common/Xml/XmlPokeAliases.cs (hunk 6 above)
Comment (RichiCoder1) [inline src/Cake.Common/Xml/XmlPokeAliases.cs]: I'll test again to be sure.

**Comment by RichiCoder1** [comment]
Updated to address comments and add examples.

**Comment by patriksvensson** [comment]
Perfect. Great work! :+1:

**Comment by patriksvensson** [comment]
Merged!

**Pull request closed without merging**